[FEAT/#292] 구글 로그인 구현합니다.

MoonsuKang opened the pull request:
## 📌 ISSUE
<!--이슈 번호 및 제목을 적어주세요!-->
closed #292 

## 📄 Work Description
<!--어떤 작업을 했는지 작성해주세요!-->
구글 로그인 구현

## ✨ PR Point
<!--리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요!! 코드를 넣어도 됩니다!-->


## 📸 ScreenShot/Video
<!--스크린샷이나 영상을 첨부해주세요! 생략 하셔도 무방합니다!-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added Google Sign-In support for user authentication and sign-up.
  * Introduced unified OAuth provider management for Google and Kakao.
  * Implemented secure storage for OAuth tokens and platform data.
  * Added Google Sign-In helper class for streamlined authentication flow.

* **Improvements**
  * Unified and streamlined login and sign-up flows with enhanced error handling.
  * Updated splash screen logic to improve auto-login and navigation reliability.
  * Integrated Google Sign-In using Activity Result API for smoother user experience.
  * Enhanced account management UI to dynamically display OAuth platform icons.
  * Refined login and sign-up UI and state management to support OAuth provider types.

* **Bug Fixes**
  * Excluded authorization headers from Google OAuth endpoints to prevent issues.

* **Chores**
  * Added new dependencies for authentication and data storage capabilities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MoonsuKang committed:
[ADD/#292] 구글 로그인 관련 라이브러리 추가

- DataStore Preferences 라이브러리 의존성 추가

MoonsuKang committed:
[REFACTOR/#292] 스플래시 화면 자동 로그인 로직 및 버전 확인 탐색 개선

- `attemptAutoLogin` 메서드가 로그인 상태를 반환하도록 수정했습니다.

MoonsuKang committed:
[ADD/#292] OAuth 정보 저장 및 관리 기능 추가

- OAuthDataStore 추가
    - ID 토큰 및 플랫폼 정보 저장, 조회, 삭제 기능 구현
    - DataStore를 사용하여 OAuth 관련 정보 관리
- OAuthDataStoreKeys 추가
    - DataStore 키 정의

MoonsuKang committed:
[FEAT/#292] 구글 회원가입 API 연동

- AuthService에 구글 회원가입 API 함수 추가
- AuthInterceptor에 구글 회원가입 경로 추가 (헤더 미포함 처리)
- GoogleSignUpRequestDto 추가
- AuthDataSource 및 AuthDataSourceImpl에 구글 회원가입 관련 함수 추가

MoonsuKang committed:
[FEAT/#292] Google 회원가입 기능 추가

AuthRepository와 AuthRepositoryImpl에 `signUpWithGoogle` 함수를 추가

MoonsuKang committed:
[ADD/#292] 로그인 타입 분기

MoonsuKang committed:
[FEAT/#292] 구글 로그인 기능 구현

- 구글 로그인 기능을 구현했습니다.
- GoogleSignInHelper 구현
- 404면 회원가입 프로세스 진행
- 추후 회원가입 때 어떤 타입의 회원가입(구글/카카오)인지 판단하기 위해 프로파이더로 분기

MoonsuKang committed:
[FEAT/#292] 회원가입 API 연동 및 구글 회원가입 로직 추가

회원가입 API 연동
- 회원가입 API에 구글 로그인 로직 추가
- Kakao와 Google 플랫폼에 따른 회원가입 분기 처리

MoonsuKang committed:
[REFACTOR/#292] 로그인 API 요청 시 불필요한 값 제거

로그인 API 요청 시 불필요한 platform, name 값 제거
GoogleSignUpRequestDto의 platform, name 제거

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/ui/splash/SplashViewModel.kt:
```diff
@@ -76,21 +76,20 @@ class SplashViewModel @AssistedInject constructor(
+    private fun attemptAutoLogin(): Boolean {
         val isLoggedIn = tokenRepository.getAccessToken().isNotBlank() &&
             tokenRepository.getRefreshToken().isNotBlank()
         setState { copy(isUserLoggedIn = isLoggedIn) }
```
[p3]
현재 로직 수정으로 인해 SplashContract.State.isLoggedIn 부분이 필요없어진 것 같습니다. 확인부탁드립니다!

SYAAINN commented on app/src/main/java/com/sopt/clody/presentation/ui/splash/SplashViewModel.kt:
[p5]
이제 스플래시 뷰모델 로직은 
디버그/릴리즈 버전 체크 && 점검시간 체크 -> 토큰 유뮤로 자동 로그인 체크 -> 앱 버전 업데이트 체크 -> 최신 버전이고 자동 로그인 가능하면 홈 아니면 로그인 화면으로 이동이겠네요!

SYAAINN commented on app/src/main/java/com/sopt/clody/data/datastore/OAuthDataStore.kt:
[p4]
preferences 기반 토큰 저장 방식이 아닌 preferences datastore 로 사용한 것 같은데, 추후에 카카오 로그인도 이것과 합치나요? 동일하게 preferences가 아닌 해당 방식을 채택한 이유가 궁금합니다!

MoonsuKang replied:
음. `SharedPreferences`는 **Sync I/O** 기반이라 UI thread를 막을 수 있어서 구조적으로 위험할 수 잇죠.
반면, `DataStore`는 Coroutine 기반으로 설계되어 IO 안정성이 좋습니다.
그래서 공식문서에서도 장기적으로 Datastore을 지향하고 있슴다.
카카오같은 경우는 로그인/회원가입에 loginsdk를 통해서 auth token을 발급하는 구조라 구글 로그인과는 조금 다를 수 있습니다.
구글 로그인은 요청하면 또 이메일 선택 UI가 뜨기 때문에 로컬에 캐싱했다가 회원가입 때 쏘는 구조로 했슴다이.

MoonsuKang committed:
[REFACTOR/#292] 로그인 타입 로직 LanguageProvider로 이동

- LanguageProvider를 통해 로그인 타입 결정
- LanguageProviderImpl에서 시스템 언어 설정에 따라 로그인 타입 반환

MoonsuKang committed:
[FEAT/#292] 로그인 플랫폼 별 아이콘 분기 처리

- UserInfoResponseDto의 platform 타입을 OAuthProvider로 변경
- AccountManagementViewModel에서 platform Enum 처리
- AccountManagementLogoutOption에서 로그인 플랫폼에 따라 아이콘을 다르게 표시

MoonsuKang committed:
[REFACTOR/#292] SplashState에서 불필요한 isUserLoggedIn 제거
